libdnf/advisory/advisory_query.cpp

```diff
@@ -0,0 +1,437 @@
+    ifilter_reference(patterns, AdvisoryReferenceType::BUGZILLA, cmp_type);
+    return *this;
+}
+
+AdvisoryQuery & AdvisoryQuery::ifilter_severity(const std::string & pattern, sack::QueryCmp cmp_type) {
+    const std::vector<std::string> patterns = {pattern};
+    ifilter_severity(patterns, cmp_type);
+    return *this;
+}
+AdvisoryQuery & AdvisoryQuery::ifilter_severity(const std::vector<std::string> & patterns, sack::QueryCmp cmp_type) {
+    auto & solv_sack = advisory_sack->base->get_rpm_solv_sack();
+    Pool * pool = solv_sack.p_impl->get_pool();
+    libdnf::rpm::solv::SolvMap filter_result(solv_sack.p_impl->get_nsolvables());
+
+    bool cmp_not = (cmp_type & libdnf::sack::QueryCmp::NOT) == libdnf::sack::QueryCmp::NOT;
+    if (cmp_not) {
+        // Removal of NOT CmpType makes following comparissons easier and more effective
+        cmp_type = cmp_type - libdnf::sack::QueryCmp::NOT;
+    }
+
+    switch (cmp_type) {
+        case libdnf::sack::QueryCmp::EQ: {
+            for (const std::string & severity : patterns) {
+                for (Id candidate_id : p_impl->data_map) {
+                    const char * candidate_severity = pool_lookup_str(pool, candidate_id, UPDATE_SEVERITY);
+                    if (!strcmp(candidate_severity, severity.c_str())) {
+                        filter_result.add_unsafe(candidate_id);
+                    }
+                }
+            }
+        } break;
+        default:
+            throw NotSupportedCmpType("Unsupported CmpType");
+    }
+
+
+    // Apply filter results to query
+    if (cmp_not) {
+        p_impl->data_map -= filter_result;
+    } else {
+        p_impl->data_map &= filter_result;
+    }
+
+    return *this;
+}
+
+AdvisoryQuery & AdvisoryQuery::ifilter_applicability(const bool value) {
+    auto & solv_sack = advisory_sack->base->get_rpm_solv_sack();
+    libdnf::rpm::solv::SolvMap filter_result(solv_sack.p_impl->get_nsolvables());
+
+    for (Id candidate_id : p_impl->data_map) {
+        Advisory a = Advisory(solv_sack, AdvisoryId(candidate_id));
+        if (a.is_applicable()) {
+            filter_result.add_unsafe(candidate_id);
+        }
+    }
+
+    if (!value) {
+        p_impl->data_map -= filter_result;
+    } else {
+        p_impl->data_map &= filter_result;
+    }
+
+    return *this;
+}
+
+AdvisoryQuery & AdvisoryQuery::ifilter_package(const libdnf::rpm::Package & package, sack::QueryCmp cmp_type) {
+    libdnf::rpm::PackageSet ps(&(advisory_sack->base->get_rpm_solv_sack()));
+    ps.add(package);
+    ifilter_packages(ps, cmp_type);
+    return *this;
+}
+
+
+AdvisoryQuery & AdvisoryQuery::ifilter_packages(const libdnf::rpm::PackageSet & package_set, sack::QueryCmp cmp_type) {
+    auto & solv_sack = advisory_sack->base->get_rpm_solv_sack();
+    Pool * pool = solv_sack.p_impl->get_pool();
+    libdnf::rpm::solv::SolvMap filter_result(solv_sack.p_impl->get_nsolvables());
+    std::vector<AdvisoryPackage> pkgs = get_sorted_advisory_packages();
```

Comment:
In the filter we use two types of packages (rpm, abvisory). I would recommend to add type of package into the name of the variable.